libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt

```diff
@@ -457,29 +487,44 @@ class MapboxNavigation(
     }
 
     /**
-     * Attach your own controller to determine when drivers arrived at stops via [ArrivalController]
+     * Set your own controller to determine when drivers arrived at stops via [ArrivalController].
      * Use [navigateNextRouteLeg] to manually move navigator to the next stop. To reset to the
-     * automatic arrival controller, call [attachArrivalController].
+     * automatic arrival controller, call [setArrivalController].
+     *
+     * By default uses [AutoArrivalController].
+     *
+     * Set *null* disables arrival at stops completely. Use this if you want to write your
+     * own mechanism for handling arrival at stops.
      *
      * @param arrivalController [ArrivalController]
      */
     @JvmOverloads
-    fun attachArrivalController(arrivalController: ArrivalController = AutoArrivalController()) {
-        arrivalProgressObserver.attach(arrivalController)
-        tripSession.registerRouteProgressObserver(arrivalProgressObserver)
+    fun setArrivalController(arrivalController: ArrivalController? = AutoArrivalController()) {
+        if (arrivalController == null) {
+            tripSession.unregisterRouteProgressObserver(arrivalProgressObserver)
+        } else {
+            arrivalProgressObserver.attach(arrivalController)
+            tripSession.registerRouteProgressObserver(arrivalProgressObserver)
+        }
     }
 
     /**
-     * Disable arrival at stops completely. Use this if you want to write your
-     * own mechanism for handling arrival at stops.
+     * Set [RerouteController]. By default uses [MapboxRerouteController].
+     *
+     * Set *null* disables auto-reroute.
      */
-    fun removeArrivalController() {
-        tripSession.unregisterRouteProgressObserver(arrivalProgressObserver)
+    fun setRerouteController(rerouteController: RerouteController?) {
+        this.rerouteController = rerouteController
```

Comment:
```suggestion
        this.rerouteController?.interrupt()
        this.rerouteController = rerouteController
```

Reply:
@LukasPaczos in this case, if reroute process has started, we interrupt it and didn't get new route 🤔

Reply:
```
    fun setRerouteController(rerouteController: RerouteController?) {
        val legacyRerouteController = this.rerouteController
        this.rerouteController = rerouteController
        
        if (legacyRerouteController?.state == RerouteState.FetchingRoute){
            legacyRerouteController.interrupt()
            reroute()
        }
    }
``` 
wdyt?

Reply:
I think we need a Manager on top of RerouteController to handle this types of issue. It's internal kitchen, non public, so might add Manager any time

Reply:
LGTM :+1:

Reply:
It'd be great to cover this test with a test.

Reply:
already in https://github.com/mapbox/mapbox-navigation-android/pull/3184/files#diff-374b1e7c64751d9ceba5da75c13562b7R393-R411 😜